service-registry/registry-etcd/src/main/java/org/apache/servicecomb/registry/etcd/EtcdDiscovery.java

```diff
@@ -0,0 +1,207 @@
+
+  private Environment environment;
+
+  private String basePath;
+
+  private EtcdRegistryProperties etcdRegistryProperties;
+
+  private Client client;
+
+  private InstanceChangedListener<EtcdDiscoveryInstance> instanceChangedListener;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(EtcdDiscovery.class);
+
+  @Autowired
+  @SuppressWarnings("unused")
+  public void setEnvironment(Environment environment) {
+    this.environment = environment;
+  }
+
+  @Autowired
+  @SuppressWarnings("unused")
+  public void setEtcdRegistryProperties(EtcdRegistryProperties etcdRegistryProperties) {
+    this.etcdRegistryProperties = etcdRegistryProperties;
+  }
+
+  @Override
+  public String name() {
+    return EtcdConst.ETCD_REGISTRY_NAME;
+  }
+
+  @Override
+  public boolean enabled(String application, String serviceName) {
+    return environment.getProperty(String.format(EtcdConst.ETCD_DISCOVERY_ENABLED, application, serviceName),
+        boolean.class, true);
+  }
+
+  @Override
+  public List<EtcdDiscoveryInstance> findServiceInstances(String application, String serviceName) {
+
+    String prefixPath = basePath + "/" + application + "/" + serviceName;
+    SingletonManager.getInstance().computeIfAbsent(prefixPath, serName -> {
+      Watch watchClient = client.getWatchClient();
+      try {
+        ByteSequence prefixByteSeq = ByteSequence.from(prefixPath, Charset.defaultCharset());
+        watchClient.watch(
+            prefixByteSeq,
+            WatchOption.builder().withPrefix(prefixByteSeq).build(),
+            resp -> {
+              Map<String, EtcdDiscoveryInstance> instanceMap = SingletonManager.getInstance()
+                  .get("instance:" + prefixPath);
+              resp.getEvents().forEach(event -> {
+                String key = event.getKeyValue().getKey().toString(Charset.defaultCharset());
+                switch (event.getEventType()) {
+                  case PUT -> {
+                    EtcdDiscoveryInstance newInstance = getEtcdDiscoveryInstance(event.getKeyValue());
+                    instanceMap.put(key, newInstance);
+                  }
+                  case DELETE -> instanceMap.remove(key);
+                  default -> LOGGER.error("unkonw event");
+                }
+              });
+              List<EtcdDiscoveryInstance> discoveryInstanceList = new ArrayList<>(instanceMap.values());
+              instanceChangedListener.onInstanceChanged(name(), application, serviceName, discoveryInstanceList);
+            }
+        );
+      } catch (Exception e) {
+        LOGGER.error("Failed to add watch", e);
+      }
+      return watchClient;
+    });
+
+    List<KeyValue> endpointKv = getValuesByPrefix(prefixPath);
+    List<EtcdDiscoveryInstance> etcdDiscoveryInstances = convertServiceInstanceList(endpointKv);
+    SingletonManager.getInstance().computeIfAbsent("instance:" + prefixPath, v -> etcdDiscoveryInstances.stream()
```

Comment:
Do not hard code constants. See above, just define a new map. Or you can listen to all events and if changed, just query all instances.

Reply:
My first version of the code was written according to this idea. When the node changes, all instances will be queried, but an exception will be thrown that the event loop is blocked. I think vert may have some checks to prevent the event loop from blocking. The second reason is that the performance of the query is also average.I will modify some hard code.